lib/cursor/QueryCursor.js

```diff
@@ -295,40 +295,102 @@ function _next(ctx, cb) {
   }
 
   if (ctx.cursor) {
-    return ctx.cursor.next(function(error, doc) {
-      if (error) {
-        return callback(error);
-      }
-      if (!doc) {
+    if (ctx.query._mongooseOptions.populate && !ctx._pop) {
+      ctx._pop = helpers.preparePopulationOptionsMQ(ctx.query,
+        ctx.query._mongooseOptions);
+      ctx._pop.__noPromise = true;
+    }
+    if (ctx.query._mongooseOptions.populate && ctx.options.batchSize > 1) {
+      if (ctx._batchDocs && ctx._batchDocs.length) {
+        // Return a cached populated doc
+        return _nextDoc(ctx, ctx._batchDocs.shift(), ctx._pop, callback);
+      } else if (ctx._batchExhausted) {
+        // Internal cursor reported no more docs. Act the same here
         return callback(null, null);
+      } else {
+        // Request as many docs as batchSize, to populate them also in batch
+        return ctx.cursor.next(_onNext.bind({ ctx, callback, batchDocs: [] }));
       }
+    } else {
+      return ctx.cursor.next(function(error, doc) {
+        if (error) {
+          return callback(error);
+        }
+        if (!doc) {
+          return callback(null, null);
+        }
 
-      const opts = ctx.query._mongooseOptions;
-      if (!opts.populate) {
-        return opts.lean ?
-          callback(null, doc) :
-          _create(ctx, doc, null, callback);
-      }
-
-      const pop = helpers.preparePopulationOptionsMQ(ctx.query,
-        ctx.query._mongooseOptions);
-      pop.__noPromise = true;
-      ctx.query.model.populate(doc, pop, function(err, doc) {
-        if (err) {
-          return callback(err);
+        if (!ctx.query._mongooseOptions.populate) {
+          return _nextDoc(ctx, doc, null, callback);
         }
-        return opts.lean ?
-          callback(null, doc) :
-          _create(ctx, doc, pop, callback);
+
+        ctx.query.model.populate(doc, ctx._pop, function(err, doc) {
+          if (err) {
+            return callback(err);
+          }
+          return _nextDoc(ctx, doc, ctx._pop, callback);
+        });
       });
-    });
+    }
   } else {
     ctx.once('cursor', function() {
       _next(ctx, cb);
     });
   }
 }
 
+/*!
+ * ignore
+ */
+
+function _onNext(error, doc) {
+  if (error) {
+    return this.callback(error);
+  }
+  if (!doc) {
+    this.ctx._batchExhausted = true;
+    return _populateBatch.call(this);
+  }
+
+  this.batchDocs.push(doc);
+
+  if (this.batchDocs.length < this.ctx.options.batchSize) {
+    this.ctx.cursor.next(_onNext.bind(this));
+  } else {
+    _populateBatch.call(this);
+  }
+}
+
+/*!
+ * ignore
+ */
+
+function _populateBatch() {
+  if (!this.batchDocs.length) {
+    return this.callback(null, null);
+  }
+  const _this = this;
+  this.ctx.query.model.populate(this.batchDocs, this.ctx._pop, function(err, docs) {
+    if (err) {
+      return _this.callback(err);
+    }
+
+    _this.ctx._batchDocs = docs;
```

Comment:
Is there a reason to have both `.batchDocs` and `._batchDocs`? `Model.populate(docs)` modifies `docs` in place so they point to the same object.

Reply:
Note `this.batchDocs` belongs to the custom object initially bound to `_onNext`, while `this.ctx._batchDocs` belongs to the QueryCursor instance. So this assignation is basically passing the Array from one object to another.

Sorry i did not researched that deep to know that `Model.populate(docs)` modified the Array in place, so maybe i could even reset `ctx._batchDocs` just before line 312 and always Keep one Array. Let me see if that works as expected.